### ISSUE TITLE
feishin scratchpad issue

### ISSUE THREAD

**Issue by fybx** [hyprland-community/pyprland#20]
Hi! I tried to pair [feishin](https://github.com/jeffvli/feishin) with pyprland to achieve a toggleable set-up, but I couldn't get pypr to hide feishin when focus was lost. In addition to that, pypr also reports this error when launching feishin:

```
Client address is invalid
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/pyprland/command.py", line 75, in _callHandler
    await getattr(plugin, full_name)(*params)
  File "/usr/lib/python3.11/site-packages/pyprland/plugins/scratchpads.py", line 389, in run_toggle
    await self.run_show(uid)
  File "/usr/lib/python3.11/site-packages/pyprland/plugins/scratchpads.py", line 429, in run_show
    await item.updateClientInfo()
  File "/usr/lib/python3.11/site-packages/pyprland/plugins/scratchpads.py", line 194, in updateClientInfo
    client_info = await get_client_props(addr="0x" + self.address)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/pyprland/plugins/scratchpads.py", line 52, in get_client_props
    assert len(addr) > 2, "Client address is invalid"
           ^^^^^^^^^^^^^
AssertionError: Client address is invalid
```

---

Here's my `pyprland.json`:

```json
{
  "pyprland": {
    "plugins": ["scratchpads", "magnify"]
  },
  "scratchpads": {
    "terminal": {
      "command": "kitty --class scratchpad",
      "margin": 50,
      "animation": "fromBottom",
      "unfocus": "hide"
    },
    "music": {
      "command": "/usr/bin/feishin",
      "lazy": true,
      "animation": "fromBottom",
      "unfocus": "hide"
    }
  }
}
```

and relevant parts from `hyprland.conf`:

```
bind = SUPER SHIFT, M,      exec, pypr toggle music     && hyprctl dispatch bringactivetotop

$music          = class:^(feishin)$
windowrulev2    = float,                    $music 
windowrulev2    = size 80% 80%,             $music
windowrulev2    = center,                   $music
```

**Comment by Morphexe** [comment]
This seems to be a regression of sorts, I updated to latest version and this started happening. Going back to 1.4 seems to do the trick.

**Comment by Morphexe** [comment]
@fybx  seems like I was wrong:
https://github.com/hyprland-community/pyprland/wiki/Troubleshooting

I was able to get latest working with this fix/change.

**Comment by fybx** [comment]
Thanks @Morphexe for wiki link, here's what I've done:

1. launched pypr with `--debug`
2. added `"class": "feishin"`

and now, feishin straight up doesn't launch :)

---

updated `pyprland.json`:

```json
"music": {
      "command": "/usr/bin/feishin",
      "lazy": true,
      "animation": "fromBottom",
      "unfocus": "hide",
      "class": "feishin"
}
```

<details>
<summary>debug log from pypr (<b>caution</b>, very long):</summary>

```
                      ipc - Logger initialized for ipc // common.py:70
                  startup - Logger initialized for startup // common.py:70
                     pypr - Logger initialized for pypr // common.py:70
              scratchpads - Logger initialized for scratchpads // common.py:70
              scratchpads - {'terminal'} // scratchpads.py:254
              scratchpads - terminal is not running, restarting... // scratchpads.py:264
              scratchpads - starting terminal // scratchpads.py:271
              scratchpads - scratch terminal has pid 30582 // scratchpads.py:301
              scratchpads - ==> Wait for terminal spawning // scratchpads.py:273
                      ipc - clients // ipc.py:25
                      ipc - clients // ipc.py:25
              scratchpads - => terminal info received on time // scratchpads.py:279
              scratchpads - => spawned terminal as proc 30582 // scratchpads.py:283
                     pypr - EVT: activewindow(kitty,pypr --debug /dev/nu ~) // command.py:99
                     pypr - EVT: activewindowv2(563df36253a0) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df36253a0\n',) // command.py:73
                     pypr - EVT: openwindow(563df3626870,special silent,scratchpad,fish) // command.py:99
                     pypr - scratchpads.event_openwindow('563df3626870,special silent,scratchpad,fish\n',) // command.py:73
                     pypr - EVT: windowtitle(563df3626870) // command.py:99
                     pypr - CMD: run_toggle(['terminal']) // command.py:129
                     pypr - scratchpads.run_toggle('terminal',) // command.py:73
              scratchpads - terminal is visible = False // scratchpads.py:385
                      ipc - activewindow // ipc.py:25
              scratchpads - Showing terminal // scratchpads.py:427
                      ipc - clients // ipc.py:25
                      ipc - clients // ipc.py:25
                      ipc - movetoworkspacesilent special:scratch_terminal,address:0x563df3626870 // ipc.py:53
                     pypr - EVT: movewindow(563df3626870,special:scratch_terminal) // command.py:99
                     pypr - EVT: activewindow(kitty,pypr --debug /dev/nu ~) // command.py:99
                     pypr - EVT: activewindowv2(563df36253a0) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df36253a0\n',) // command.py:73
                      ipc - monitors // ipc.py:25
                      ipc - monitors // ipc.py:25
                      ipc - moveworkspacetomonitor special:scratch_terminal HDMI-A-1 // ipc.py:53
                      ipc - movetoworkspacesilent 4,address:0x563df3626870 // ipc.py:53
                     pypr - EVT: moveworkspace(special:scratch_terminal,HDMI-A-1) // command.py:99
                     pypr - EVT: movewindow(563df3626870,4) // command.py:99
                     pypr - EVT: activewindow(scratchpad,~) // command.py:99
                     pypr - EVT: activewindowv2(563df3626870) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3626870\n',) // command.py:73
                      ipc - movewindowpixel exact 2304 372,address:0x563df3626870 // ipc.py:53
                      ipc - focuswindow address:0x563df3626870 // ipc.py:53
                     pypr - EVT: activewindow(scratchpad,~) // command.py:99
                     pypr - EVT: activewindowv2(563df3626870) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3626870\n',) // command.py:73
                     pypr - EVT: activewindow(kitty,pypr --debug /dev/nu ~) // command.py:99
                     pypr - EVT: activewindowv2(563df36253a0) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df36253a0\n',) // command.py:73
              scratchpads - hide terminal because another client is active // scratchpads.py:337
              scratchpads - Hiding terminal // scratchpads.py:468
                      ipc - movewindowpixel 0 842,address:0x563df3626870 // ipc.py:53
                      ipc - movetoworkspacesilent special:scratch_terminal,address:0x563df3626870 // ipc.py:53
                     pypr - EVT: movewindow(563df3626870,special:scratch_terminal) // command.py:99
                     pypr - EVT: windowtitle(563df36253a0) // command.py:99
                     pypr - EVT: activewindow(kitty,fish) // command.py:99
                     pypr - EVT: activewindowv2(563df36253a0) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df36253a0\n',) // command.py:73
                     pypr - EVT: windowtitle(563df36253a0) // command.py:99
                     pypr - EVT: activewindow(kitty,~) // command.py:99
                     pypr - EVT: activewindowv2(563df36253a0) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df36253a0\n',) // command.py:73
                     pypr - CMD: run_toggle(['terminal']) // command.py:129
                     pypr - scratchpads.run_toggle('terminal',) // command.py:73
              scratchpads - terminal is visible = False // scratchpads.py:385
                      ipc - activewindow // ipc.py:25
              scratchpads - Showing terminal // scratchpads.py:427
                      ipc - clients // ipc.py:25
                      ipc - monitors // ipc.py:25
                      ipc - moveworkspacetomonitor special:scratch_terminal HDMI-A-1 // ipc.py:53
                      ipc - movetoworkspacesilent 4,address:0x563df3626870 // ipc.py:53
                     pypr - EVT: movewindow(563df3626870,4) // command.py:99
                      ipc - movewindowpixel exact 2304 372,address:0x563df3626870 // ipc.py:53
                      ipc - focuswindow address:0x563df3626870 // ipc.py:53
                     pypr - EVT: activewindow(scratchpad,~) // command.py:99
                     pypr - EVT: activewindowv2(563df3626870) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3626870\n',) // command.py:73
                     pypr - EVT: windowtitle(563df3626870) // command.py:99
                     pypr - EVT: activewindow(scratchpad,vim .config/hypr/pyp ~) // command.py:99
                     pypr - EVT: activewindowv2(563df3626870) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3626870\n',) // command.py:73
                     pypr - EVT: activewindow(chromium,Troubleshooting · hyprland-community/pyprland Wiki - Chromium) // command.py:99
                     pypr - EVT: activewindowv2(563df429e900) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df429e900\n',) // command.py:73
              scratchpads - hide terminal because another client is active // scratchpads.py:337
              scratchpads - Hiding terminal // scratchpads.py:468
                      ipc - movewindowpixel 0 842,address:0x563df3626870 // ipc.py:53
                      ipc - movetoworkspacesilent special:scratch_terminal,address:0x563df3626870 // ipc.py:53
                     pypr - EVT: workspace(2) // command.py:99
                     pypr - EVT: movewindow(563df3626870,special:scratch_terminal) // command.py:99
                     pypr - CMD: run_toggle(['terminal']) // command.py:129
                     pypr - scratchpads.run_toggle('terminal',) // command.py:73
              scratchpads - terminal is visible = False // scratchpads.py:385
                      ipc - activewindow // ipc.py:25
              scratchpads - Showing terminal // scratchpads.py:427
                      ipc - clients // ipc.py:25
                      ipc - monitors // ipc.py:25
                      ipc - moveworkspacetomonitor special:scratch_terminal HDMI-A-1 // ipc.py:53
                      ipc - movetoworkspacesilent 2,address:0x563df3626870 // ipc.py:53
                      ipc - movewindowpixel exact 2304 372,address:0x563df3626870 // ipc.py:53
                     pypr - EVT: movewindow(563df3626870,2) // command.py:99
                      ipc - focuswindow address:0x563df3626870 // ipc.py:53
                     pypr - EVT: activewindow(scratchpad,vim .config/hypr/pyp ~) // command.py:99
                     pypr - EVT: activewindowv2(563df3626870) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3626870\n',) // command.py:73
                     pypr - EVT: windowtitle(563df3626870) // command.py:99
                     pypr - EVT: activewindow(scratchpad,~) // command.py:99
                     pypr - EVT: activewindowv2(563df3626870) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3626870\n',) // command.py:73
                     pypr - EVT: activewindow(obsidian,digital-commonplace - 400-499 vault - Obsidian v1.4.16) // command.py:99
                     pypr - EVT: activewindowv2(563df3622a00) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3622a00\n',) // command.py:73
              scratchpads - hide terminal because another client is active // scratchpads.py:337
              scratchpads - Hiding terminal // scratchpads.py:468
                      ipc - movewindowpixel 0 842,address:0x563df3626870 // ipc.py:53
                      ipc - movetoworkspacesilent special:scratch_terminal,address:0x563df3626870 // ipc.py:53
                     pypr - EVT: workspace(3) // command.py:99
                     pypr - EVT: movewindow(563df3626870,special:scratch_terminal) // command.py:99
                     pypr - EVT: activewindow(kitty,~) // command.py:99
                     pypr - EVT: activewindowv2(563df36253a0) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df36253a0\n',) // command.py:73
                     pypr - EVT: workspace(4) // command.py:99
                     pypr - EVT: windowtitle(563df36253a0) // command.py:99
                     pypr - EVT: activewindow(kitty,pypr reload ~) // command.py:99
                     pypr - EVT: activewindowv2(563df36253a0) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df36253a0\n',) // command.py:73
                     pypr - CMD: run_reload([]) // command.py:129
                     pypr - builtin.run_reload() // command.py:73
              scratchpads - set() // scratchpads.py:254
                     pypr - EVT: windowtitle(563df36253a0) // command.py:99
                     pypr - EVT: activewindow(kitty,~) // command.py:99
                     pypr - EVT: activewindowv2(563df36253a0) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df36253a0\n',) // command.py:73
                     pypr - CMD: run_toggle(['music']) // command.py:129
                     pypr - scratchpads.run_toggle('music',) // command.py:73
              scratchpads - music is visible = False // scratchpads.py:385
                      ipc - activewindow // ipc.py:25
              scratchpads - Showing music // scratchpads.py:427
              scratchpads - music is not running, restarting... // scratchpads.py:264
              scratchpads - starting music // scratchpads.py:271
              scratchpads - scratch music has pid 30827 // scratchpads.py:301
              scratchpads - ==> Wait for music spawning // scratchpads.py:273
                      ipc - clients // ipc.py:25
                      ipc - clients // ipc.py:25
                      ipc - clients // ipc.py:25
                      ipc - clients // ipc.py:25
                      ipc - clients // ipc.py:25
                      ipc - clients // ipc.py:25
                      ipc - clients // ipc.py:25
                      ipc - clients // ipc.py:25
                      ipc - clients // ipc.py:25
                      ipc - clients // ipc.py:25
              scratchpads - => spawned music as proc 30827 // scratchpads.py:283
                     pypr - Bug detected, please report on https://github.com/fdev31/pyprland/issues // command.py:77
                     pypr - Client address is invalid // command.py:80
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/pyprland/command.py", line 75, in _callHandler
    await getattr(plugin, full_name)(*params)
  File "/usr/lib/python3.11/site-packages/pyprland/plugins/scratchpads.py", line 389, in run_toggle
    await self.run_show(uid)
  File "/usr/lib/python3.11/site-packages/pyprland/plugins/scratchpads.py", line 429, in run_show
    await item.updateClientInfo()
  File "/usr/lib/python3.11/site-packages/pyprland/plugins/scratchpads.py", line 194, in updateClientInfo
    client_info = await get_client_props(addr="0x" + self.address)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/pyprland/plugins/scratchpads.py", line 52, in get_client_props
    assert len(addr) > 2, "Client address is invalid"
           ^^^^^^^^^^^^^
AssertionError: Client address is invalid
                     pypr - EVT: activewindow(feishin,Feishin) // command.py:99
                     pypr - EVT: activewindowv2(563df42fe500) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df42fe500\n',) // command.py:73
                     pypr - EVT: openwindow(563df42fe500,4,feishin,Feishin) // command.py:99
                     pypr - scratchpads.event_openwindow('563df42fe500,4,feishin,Feishin\n',) // command.py:73
              scratchpads - Lookup hack triggered // scratchpads.py:349
                      ipc - clients // ipc.py:25
              scratchpads - client class found: {'address': '0x563df42fe500', 'mapped': True, 'hidden': False, 'at': [2112, 108], 'size': [1536, 864], 'workspace': {'id': 4, 'name': '4'}, 'floating': True, 'monitor': 1, 'class': 'feishin', 'title': 'Feishin', 'initialClass': 'feishin', 'initialTitle': 'Feishin', 'pid': 30827, 'xwayland': True, 'pinned': False, 'fullscreen': False, 'fullscreenMode': 0, 'fakeFullscreen': False, 'grouped': [], 'swallowing': '0x0'} // scratchpads.py:356
                      ipc - clients // ipc.py:25
                      ipc - movetoworkspacesilent special:scratch_music,address:0x563df42fe500 // ipc.py:53
                      ipc - monitors // ipc.py:25
                     pypr - EVT: createworkspace(special:scratch_music) // command.py:99
                     pypr - EVT: movewindow(563df42fe500,special:scratch_music) // command.py:99
                     pypr - EVT: activewindow(kitty,~) // command.py:99
                     pypr - EVT: activewindowv2(563df36253a0) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df36253a0\n',) // command.py:73
                     pypr - EVT: activewindow(chromium,Troubleshooting · hyprland-community/pyprland Wiki - Chromium) // command.py:99
                     pypr - EVT: activewindowv2(563df429e900) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df429e900\n',) // command.py:73
                     pypr - EVT: workspace(2) // command.py:99
                     pypr - EVT: activewindow(obsidian,digital-commonplace - 400-499 vault - Obsidian v1.4.16) // command.py:99
                     pypr - EVT: activewindowv2(563df3622a00) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3622a00\n',) // command.py:73
                     pypr - EVT: workspace(3) // command.py:99
                     pypr - EVT: activewindow(kitty,~) // command.py:99
                     pypr - EVT: activewindowv2(563df36253a0) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df36253a0\n',) // command.py:73
                     pypr - EVT: workspace(4) // command.py:99
                     pypr - CMD: run_toggle(['terminal']) // command.py:129
                     pypr - scratchpads.run_toggle('terminal',) // command.py:73
              scratchpads - terminal is visible = False // scratchpads.py:385
                      ipc - activewindow // ipc.py:25
              scratchpads - Showing terminal // scratchpads.py:427
                      ipc - clients // ipc.py:25
                      ipc - monitors // ipc.py:25
                      ipc - moveworkspacetomonitor special:scratch_terminal HDMI-A-1 // ipc.py:53
                      ipc - movetoworkspacesilent 4,address:0x563df3626870 // ipc.py:53
                     pypr - EVT: movewindow(563df3626870,4) // command.py:99
                      ipc - movewindowpixel exact 2304 372,address:0x563df3626870 // ipc.py:53
                      ipc - focuswindow address:0x563df3626870 // ipc.py:53
                     pypr - EVT: activewindow(scratchpad,~) // command.py:99
                     pypr - EVT: activewindowv2(563df3626870) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3626870\n',) // command.py:73
                     pypr - EVT: windowtitle(563df3626870) // command.py:99
                     pypr - EVT: activewindow(scratchpad,vim .config/hypr/hyp ~) // command.py:99
                     pypr - EVT: activewindowv2(563df3626870) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3626870\n',) // command.py:73
                     pypr - EVT: windowtitle(563df3626870) // command.py:99
                     pypr - EVT: activewindow(scratchpad,~) // command.py:99
                     pypr - EVT: activewindowv2(563df3626870) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3626870\n',) // command.py:73
                     pypr - EVT: windowtitle(563df3626870) // command.py:99
                     pypr - EVT: activewindowv2(563df3626870) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3626870\n',) // command.py:73
                     pypr - EVT: activewindow(scratchpad,vim .config/hypr/pyp ~) // command.py:99
                     pypr - EVT: windowtitle(563df3626870) // command.py:99
                     pypr - EVT: activewindowv2(563df3626870) // command.py:99
                     pypr - scratchpads.event_activewindowv2('563df3626870\n',) // command.py:73
                     pypr - EVT: activewindow(scratchpad,~) // command.py:99
```
</details>

**Comment by cunbidun** [comment]
It seems like for some slow start applications, 10 retries on line 275 are not enough:
https://github.com/hyprland-community/pyprland/blob/91e9d84924c7e205d9e259351454bf008b3283c0/pyprland/plugins/scratchpads.py#L274-L280

**Comment by fybx** [comment]
A slowly rising wait interval for 10 times might be useful. I first though of just waiting forever until a value was returned, but it's way too risky. 

```python
loop_count = count()
acc = 1
while next(loop_count) < 10: 
    await asyncio.sleep(0.1 * acc ** 2)
    acc += 1
```

**Comment by fdev31** [comment]
Whow, great group effort!
I'll push a version implementing this suggestion, this limit of ~30s total sounds already a lot, I'll make it ~14s... the new version should still let 3x more time for the app to start, be aware only one command can run at a time at the moment to keep the code simple... 
Keep me updated :)

**Comment by fdev31** [comment]
Do you have a chance to test the latest git version or should I push a new release for convenience (I'ld like to do so only after positive feedback) ?

**Comment by fdev31** [comment]
I was thinking about the failing case which is not handled today.
The process should probably be killed (+ relevant error handling) so next time there is a retry... does it sound like a good idea to you and do you have suggestions?

**Comment by fybx** [comment]
> Whow, great group effort!

Actually, I edited the code to open a PR but wasn't able to run it :( I would like to test the latest changes in the repo if you can help me :3

---

Here's what I'm doing:

In pyprland/pyprland: ```python command.py```
Gives this error:
Traceback (most recent call last):
  File "/home/ferit/shoka/300-399 repos/399 archive/pyprland/pyprland/command.py", line 11, in <module>
    from .common import PyprError, get_logger, init_logger
ImportError: attempted relative import with no known parent package

**Comment by fybx** [comment]
In addition to my previous comment, I reverted back to Sonixd instead of Feishin, as it's openly in development. I've used the same pyprland and hyprland configuration as for the Feishin; however, Sonixd and pypr worked fine this time.

I'm guessing the method used to spawn windows and child workers for Feishin has issues causing incompatibilities. We can either close this issue, or keep working on it for future's sake 🐞

**Comment by cunbidun** [comment]
Tested the latest master (3cee7c5). It seems to fix the slow start problem :)

**Comment by fdev31** [comment]
> I would like to test the latest changes in the repo if you can help me :3

The cleanest & simplest is to run it from a virtualenv:

```
python -m venv <temporary path>
<temporary path>/bin/pip install git+https://github.com/fdev31/pyprland
<temporary path>/bin/pypr --debug /tmp/logfile
```

**Comment by fdev31** [comment]
Thank you all again, I'm closing this. I released 1.4.3 including those changes.
